### PR TITLE
Generator

### PR DESCRIPTION
**Pull request by emailweixu**
Generator generates outputs given `inputs` (can be None) by transforming
a random noise and input using `net`:
    outputs = net([noise, inputs]) if input is not None
              else net(noise)

It has two modes:
ML: net is trained to minized loss_func([outputs, inputs])
STEIN: net is trained to generate outputs to match the distribution whose
    (unnormalized) negative probability is given by loss_func([outputs, inputs]).
    The matching is achieved by using amortized Stein variational gradient
    descent (SVGD). See the following paper for reference
    Feng et al "Learning to Draw Samples with Amortized Stein Variational
    Gradient Descent" https://arxiv.org/pdf/1707.06626.pdf

It also supports an additional optional objective of maximizing the mutual
information between [inputs, noise] and outputs by using mi_estimator

### CODE REVIEW

File: alf/algorithms/generator_test.py [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+        After training, w^T w is the variance of the distribution implied by the
+        generator. So it should be diag(1,4) for STEIN and 0 for 'ML'.
+        """
+        logging.info("mode: %s mi_weight: %s" % (mode, mi_weight))
+        dim = 2
+        batch_size = 512
+        net = Net(dim)
+        generator = Generator(
+            dim,
+            noise_dim=dim,
```
Comment (hnyu) [inline alf/algorithms/generator_test.py]: add test case where out_dim != noise_dim to check the bug

File: alf/algorithms/generator.py [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+            loss_propagated += self._mi_weight * mi_loss
+
+        return AlgorithmStep(
+            outputs=outputs,
+            state=(),
+            info=LossInfo(
+                loss=loss_propagated,
+                extra=GeneratorLossInfo(generator=loss, mi_estimator=mi_loss)))
+
+    def _ml_grad(self, inputs, outputs, loss_func):
```
[the remark this answers is not in the document]
Comment (emailweixu) [reply alf/algorithms/generator.py]: Because we don't want to change the parameters in loss_func, we have to do the gradient this way.

File: alf/algorithms/generator.py [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+            tape.watch(outputs)
+            loss_inputs = outputs if inputs is None else [outputs, inputs]
+            loss = loss_func(loss_inputs)
+            scalar_loss = tf.reduce_sum(loss)
+        grad = tape.gradient(scalar_loss, outputs)
+        return loss, grad
+
+    def _kernel_func(self, x, y):
+        d = x - y
+        d = d * d
```
[the remark this answers is not in the document]
Comment (emailweixu) [reply alf/algorithms/generator.py]: nice

File: alf/algorithms/generator.py [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+                info: LossInfo
+        """
+        outputs, gen_inputs = self._predict(inputs, batch_size)
+        loss, grad = self._grad_func(inputs, outputs, loss_func)
+        loss_propagated = tf.reduce_sum(
+            tf.stop_gradient(grad) * outputs, axis=-1)
+
+        mi_loss = ()
+        if self._mi_estimator is not None:
+            mi_step = self._mi_estimator.train_step([outputs, gen_inputs])
```
[the remark this answers is not in the document]
Comment (emailweixu) [reply alf/algorithms/generator.py]: good catch

File: alf/algorithms/generator.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (hnyu) [reply alf/algorithms/generator.py]: Do you mean that the parameters of loss_func will never be optimized? Or they will be optimized somewhere else in the algorithm? If it's the former case, then probably we just never watch its parameters in the code.

[the remark this answers is not in the document]
Comment (hnyu) [reply alf/algorithms/generator.py]: Do you mean that the parameters of loss_func will never be optimized? Or they will be optimized somewhere else in the algorithm? If it's the former case, then probably we just never watch its parameters in the code. Some examples of what a parameterized loss_func looks like and how its parameters are optimized will be helpful.

[the remark this answers is not in the document]
Comment (emailweixu) [reply alf/algorithms/generator.py]: loss_func is optimized somewhere else. GAN is a such example.

File: alf/algorithms/generator_test.py (hunk 1 above)
Comment (emailweixu) [reply alf/algorithms/generator_test.py]: done
